vql/server/humio/humio.go

```diff
@@ -0,0 +1,707 @@
+	self.endpointUrl = baseUrl + apiEndpoint
+	self.authToken = authToken
+
+	if baseUrl == "" {
+		return errInvalidArgument{
+			Arg: "baseUrl",
+			Err: errors.New("invalid value - must not be empty"),
+		}
+	}
+
+	_, err := url.ParseRequestURI(self.endpointUrl)
+	if err != nil {
+		return errInvalidArgument{
+			Arg: "baseUrl",
+			Err: err,
+		}
+	}
+
+	if authToken == "" {
+		return errInvalidArgument{
+			Arg: "authToken",
+			Err: errors.New("invalid value - must not be empty"),
+		}
+	}
+
+	// We want to do our own cleanup, which has a pipeline we want to flush if we can
+	self.workerGrp, self.workerCtx = errgroup.WithContext(context.Background())
+
+	self.httpClient = &http.Client{Timeout:   self.httpClientTimeoutDuration,
+				       Transport: getHttpTransport(), }
+
+	self.id = int(atomic.AddInt32(&gNextId, 1))
+	self.logPrefix = fmt.Sprintf("humio/%v: ", self.id)
+
+	options := api.QueueOptions{
+		DisableFileBuffering: false,
+		FileBufferLeaseSize: 100,
+		OwnerName: "humio-plugin",
+	}
+	self.listener, err = directory.NewListener(self.config, self.workerCtx,
+						   options.OwnerName, options)
+	if err != nil {
+		return err
+	}
+
+	self.queueOpened = 1
+	for i := 0; i < self.nWorkers; i++ {
+		self.workerGrp.Go(func() error {
+			return self.processEvents(self.workerCtx, scope)
+		})
+	}
+
+	return nil
+}
+
+func (self *HumioQueue) Opened() bool {
+	return atomic.LoadInt32(&self.queueOpened) > 0
+}
+
+func (self *HumioQueue) addDebugCallback(count int, callback func(int)) error {
+	if self.Opened() {
+		return errQueueOpened
+	}
+
+	if !self.debugEventsEnabled {
+		self.debugEventsMap = map[int][]func(int){}
+		self.debugEventsEnabled = true
+	}
+
+	_, ok := self.debugEventsMap[count]
+	if ok {
+		self.debugEventsMap[count] = append(self.debugEventsMap[count], callback)
+	} else {
+		self.debugEventsMap[count] = []func(int){callback}
+	}
+
+	return nil
+}
+
+// Provide the hostname for the client host if it's a client query
+// since an external system will not have a way to map it to a hostname.
+func (self *HumioQueue) addClientInfo(ctx context.Context, row *ordereddict.Dict,
+				      payload *HumioPayload) {
+	client_id, ok := row.GetString("ClientId")
+	if ok {
+		payload.Tags["ClientId"] = client_id
+
+		hostname := services.GetHostname(ctx, self.config, client_id)
+		if hostname != "" {
+			payload.Tags["ClientHostname"] = hostname
+		}
+	}
+}
+
+func (self *HumioQueue) addMappedTags(row *ordereddict.Dict, payload *HumioPayload) {
+	for name, mappedName := range self.tagMap {
+		var value interface{}
+		value, ok := row.Get(name)
+
+		if ok {
+			payload.Tags[mappedName] = value
+		}
+	}
+}
+
+// We don't want to parse the timestamp.  We don't actually care.  Humio can
+// handle the parsing.
+func (self *HumioQueue) addTimestamp(row *ordereddict.Dict, payload *HumioPayload) {
+	timestamp, ok := row.Get("Time")
+	timezone := ""
+	if !ok {
+		timestamp, ok = row.Get("timestamp")
```

Comment:
You can use GetInt64, GetString etc . For times you can use TimeFromAny https://github.com/Velocidex/velociraptor/blob/0f445d79cc040e6ada5c6d588a67211e9c1407b4/vql/functions/time.go#L197 to just do the right thing

I am not sure if timezone is needed? Normally times are sent in ISO format so the timezone is already in them.

Reply:
For $reasons, Humio requires a timezone when unix-epoch style timestamps are used. Agreed that this would be irrelevant if I used TimeFromAny instead and just made everything UTC.

Reply:
Or not. I'll just let the json marshaling handle the timezones